pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+package config
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	helperclient "github.com/docker/docker-credential-helpers/client"
+	"github.com/docker/docker-credential-helpers/credentials"
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/pkg/errors"
+)
+
+type dockerAuthConfig struct {
+	Auth string `json:"auth,omitempty"`
+}
+
+type dockerConfigFile struct {
+	AuthConfigs map[string]dockerAuthConfig `json:"auths"`
+	CredHelpers map[string]string           `json:"credHelpers,omitempty"`
+}
+
+const (
+	dockerHostname = "docker.io"
+	dockerRegistry = "registry-1.docker.io"
+
+	authCfg           = ".config/containers"
+	authCfgFileName   = "auth.json"
+	dockerCfg         = ".docker"
+	dockerCfgFileName = "config.json"
+
+	addAuth    = "add"
+	deleteAuth = "delete"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the username and password in the auth.json file
+func SetAuthentication(registry, username, password string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, username, password, path, addAuth)
+}
+
+// GetAuthentication returns the registry credentials stored in
+// either auth.json file or .docker/config.json
+func GetAuthentication(registry string) (string, string, error) {
+	registry = DockerHostNameToRegistry(registry)
+	// Check auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	username, password, err := findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if username != "" && password != "" {
+		return username, password, nil
+	}
+
+	// check ~/.docker/config.json as secondary option
+	path = filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName)
+	username, password, err = findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	return username, password, nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, "", "", path, deleteAuth)
+}
+
+// DockerHostNameToRegistry converts the docker hostname to the docker registry
+func DockerHostNameToRegistry(registry string) string {
```

Comment:
I’m honestly not sure that it makes sense to make this a public function: the other functions in this package do the conversion implicitly.  `docker_client.go` does need to make a similar mapping, but it’s not _entirely_ obvious that the mapping for the config file, and the mapping for the HTTP access, are always the same (e.g. see how `normalizeRegistry` below actually ends up using `index.docker.io`).

Maybe it would be better to make this mapping private in this package, and have `docker_client.go` have a separate copy, just because it’s difficult to think of a really descriptive name for this mapping function. I could argue that it should be called `DockerRegistryToHostName` instead, and neither is actually all that informative (both names are registries, both names are host names). Yes, that’s a horribly stupid reason 🤔

Reply:
fixed